Add Check Support for Java 21 Pattern Matching for Switch Syntax:  CyclomaticComplexity

child of https://github.com/checkstyle/checkstyle/issues/14961

I have read check documentation:https://checkstyle.org/checks/metrics/cyclomaticcomplexity.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

---

```
➜  src cat config.xml                               
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="CyclomaticComplexity">
            <property name="max" value="3"/>
        </module>
    </module>
</module>
➜  src cat Test.java                                
import java.awt.color.ICC_ColorSpace;

record ColoredPoint(int p, int x, boolean c) {
}
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) {
}

public class Test {

    // violation,  Cyclomatic Complexity is 4
    void test(Object obj) {  // 1, function declaration
        switch (obj) {
            case ColoredPoint(int a, int b, _)  // 2 , case
                    when (a >= b) -> {        // should we add 1 for when ? it was literally a `&&` before
            }
            case ColoredPoint(int a, int b, _) -> {}  // 3, case
            case Rectangle(_, _) -> {}  // 4, case
            default -> System.out.println("none");
        }
    }

    // violation,  Cyclomatic Complexity is 5
    void test2(Object obj) {  // 1, function declaration
        // I interpreted `when` as `&&`
        if (obj instanceof ColoredPoint(int a, int b, _) && a >= b) { // 2, if and 3, '&&'

        } else if (obj instanceof ColoredPoint(int a, int b, _)) { // 4 if

        } else if (obj instanceof Rectangle(_, _)) {   // 5 if

        }
    }
}
➜  src javac --enable-preview --release 21 Test.java
Note: Test.java uses preview features of Java SE 21.
Note: Recompile with -Xlint:preview for details.
➜  src java -jar checkstyle-10.18.1-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:11:5: Cyclomatic Complexity is 4 (max allowed is 3). [CyclomaticComplexity]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:23:5: Cyclomatic Complexity is 5 (max allowed is 3). [CyclomaticComplexity]
Audit done.
Checkstyle ends with 2 errors.

```

---
Describe what you want in detail

Since we add 1 for `&&`. I think it makes sense that `when` should be considered as +1 here. It is just the same as `&&` in if conditions see the similarities between the two examples

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/CyclomaticComplexityCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/CyclomaticComplexityCheck.java
@@ -39,10 +39,20 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  * </p>
  * <p>
  * The complexity is equal to the number of decision points {@code + 1}.
- * Decision points: {@code if}, {@code while}, {@code do}, {@code for},
- * {@code ?:}, {@code catch}, {@code switch}, {@code case} statements and
- * operators {@code &amp;&amp;} and {@code ||} in the body of target.
+ * Decision points:
  * </p>
+ * <ul>
+ * <li>
+ * {@code if}, {@code while}, {@code do}, {@code for},
+ * {@code ?:}, {@code catch}, {@code switch}, {@code case} statements.
+ * </li>
+ * <li>
+ *  Operators {@code &amp;&amp;} and {@code ||} in the body of target.
+ * </li>
+ * <li>
+ *  {@code when} expression in case labels, also known as guards.
+ * </li>
+ * </ul>
  * <p>
  * By pure theory level 1-4 is considered easy to test, 5-7 OK, 8-10 consider
  * re-factoring to ease testing, and 11+ re-factor now as testing will be painful.
@@ -93,7 +103,9 @@ import com.puppycrawl.tools.checkstyle.api.TokenTypes;
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LAND">
  * LAND</a>,
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LOR">
- * LOR</a>.
+ * LOR</a>,
+ * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_WHEN">
+ * LITERAL_WHEN</a>.
  * </li>
  * </ul>
  * <p>
@@ -177,6 +189,7 @@ public class CyclomaticComplexityCheck
             TokenTypes.LAND,
             TokenTypes.LOR,
             TokenTypes.COMPACT_CTOR_DEF,
+            TokenTypes.LITERAL_WHEN,
         };
     }
 
@@ -198,6 +211,7 @@ public class CyclomaticComplexityCheck
             TokenTypes.LAND,
             TokenTypes.LOR,
             TokenTypes.COMPACT_CTOR_DEF,
+            TokenTypes.LITERAL_WHEN,
         };
     }
 

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/metrics/CyclomaticComplexityCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/metrics/CyclomaticComplexityCheck.xml
@@ -14,10 +14,20 @@
  &lt;/p&gt;
  &lt;p&gt;
  The complexity is equal to the number of decision points {@code + 1}.
- Decision points: {@code if}, {@code while}, {@code do}, {@code for},
- {@code ?:}, {@code catch}, {@code switch}, {@code case} statements and
- operators {@code &amp;amp;&amp;amp;} and {@code ||} in the body of target.
+ Decision points:
  &lt;/p&gt;
+ &lt;ul&gt;
+ &lt;li&gt;
+ {@code if}, {@code while}, {@code do}, {@code for},
+ {@code ?:}, {@code catch}, {@code switch}, {@code case} statements.
+ &lt;/li&gt;
+ &lt;li&gt;
+  Operators {@code &amp;amp;&amp;amp;} and {@code ||} in the body of target.
+ &lt;/li&gt;
+ &lt;li&gt;
+  {@code when} expression in case labels, also known as guards.
+ &lt;/li&gt;
+ &lt;/ul&gt;
  &lt;p&gt;
  By pure theory level 1-4 is considered easy to test, 5-7 OK, 8-10 consider
  re-factoring to ease testing, and 11+ re-factor now as testing will be painful.
@@ -42,7 +52,7 @@
                <description>Control whether to treat
  the whole switch block as a single decision point.</description>
             </property>
-            <property default-value="LITERAL_WHILE,LITERAL_DO,LITERAL_FOR,LITERAL_IF,LITERAL_SWITCH,LITERAL_CASE,LITERAL_CATCH,QUESTION,LAND,LOR"
+            <property default-value="LITERAL_WHILE,LITERAL_DO,LITERAL_FOR,LITERAL_IF,LITERAL_SWITCH,LITERAL_CASE,LITERAL_CATCH,QUESTION,LAND,LOR,LITERAL_WHEN"
                       name="tokens"
                       type="java.lang.String[]"
                       validation-type="tokenSet">

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/CyclomaticComplexityCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/CyclomaticComplexityCheckTest.java
@@ -125,6 +125,7 @@ public class CyclomaticComplexityCheckTest
             TokenTypes.LAND,
             TokenTypes.LOR,
             TokenTypes.COMPACT_CTOR_DEF,
+            TokenTypes.LITERAL_WHEN,
         };
         assertWithMessage("Invalid acceptable tokens")
             .that(actual)
@@ -164,6 +165,30 @@ public class CyclomaticComplexityCheckTest
 
         verifyWithInlineConfigParser(
                 getPath("InputCyclomaticComplexitySwitchBlocks5.java"), expected);
+    }
+
+    @Test
+    public void testWhenExpression() throws Exception {
+        final String[] expected = {
+            "13:4: " + getCheckMessage(MSG_KEY, 5, 0),
+            "19:4: " + getCheckMessage(MSG_KEY, 5, 0),
+            "28:4: " + getCheckMessage(MSG_KEY, 7, 0),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputCyclomaticComplexityWhenExpression.java"), expected);
+    }
+
+    @Test
+    public void testWhenExpressionSwitchAsSinglePoint() throws Exception {
+        final String[] expected = {
+            "14:5: " + getCheckMessage(MSG_KEY, 5, 0),
+            "20:5: " + getCheckMessage(MSG_KEY, 4, 0),
+            "29:5: " + getCheckMessage(MSG_KEY, 5, 0),
+            "39:5: " + getCheckMessage(MSG_KEY, 5, 0),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                        "InputCyclomaticComplexityWhenSwitchAsSinglePoint.java"), expected);
     }
 
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/cyclomaticcomplexity/InputCyclomaticComplexityWhenExpression.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/cyclomaticcomplexity/InputCyclomaticComplexityWhenExpression.java
@@ -1,0 +1,38 @@
+/*
+CyclomaticComplexity
+max = 0
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.metrics.cyclomaticcomplexity;
+
+public class InputCyclomaticComplexityWhenExpression {
+
+   // violation below, 'Cyclomatic Complexity is 5 (max allowed is 0)'
+   void testIf(Object o) {   // 1, function declaration
+      if (o instanceof Integer i && i == 9) {}  // 2, if - 3, &&
+      else if (o instanceof String s && s.length() == 9) {} // 4, if - 5, &&
+   }
+
+   // violation below, 'Cyclomatic Complexity is 5 (max allowed is 0)'
+   void testSwitch(Object o) {   // 1, function declaration
+      switch (o) {
+         case Integer i when i == 9 -> {} // 2, case - 3, when
+         case String s when s.length() == 9 -> {} // 4, case - 5, when
+         default -> {}
+      }
+   }
+
+   // violation below, 'Cyclomatic Complexity is 7 (max allowed is 0)'
+   void testSwitch2(Object o) {  // 1, function declaration
+     switch (o) {
+         case Integer i -> {}  // 2, case
+         case Point(int x, int y) when (x == 0 && y == 0) -> {} // 3, case - 4, when - 5, &&
+         case Point(int x, int y) when x == 0 -> {} // 6, case - 7, when
+         default ->  {}
+      }
+   }
+
+   record Point(int x, int y) {}
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/cyclomaticcomplexity/InputCyclomaticComplexityWhenSwitchAsSinglePoint.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/cyclomaticcomplexity/InputCyclomaticComplexityWhenSwitchAsSinglePoint.java
@@ -1,0 +1,59 @@
+/*
+CyclomaticComplexity
+max = 0
+switchBlockAsSingleDecisionPoint = true
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.metrics.cyclomaticcomplexity;
+
+public class InputCyclomaticComplexityWhenSwitchAsSinglePoint {
+
+    // violation below, 'Cyclomatic Complexity is 5 (max allowed is 0)'
+    void testIf(Object o) {   // 1, function declaration
+        if (o instanceof Integer i && i == 9) {}  // 2, if - 3, &&
+        else if (o instanceof String s && s.length() == 9) {} // 4, if - 5, &&
+    }
+
+    // violation below, 'Cyclomatic Complexity is 4 (max allowed is 0)'
+    void testSwitch(Object o) {   // 1, function declaration
+        switch (o) {  // 2, switch
+            case Integer i when i == 9 -> {} // 3, when
+            case String s when s.length() == 9 -> {} // 4, when
+            default -> {}
+        }
+    }
+
+    // violation below, 'Cyclomatic Complexity is 5 (max allowed is 0)'
+    void testSwitch2(Object o) {  // 1, function declaration
+        switch (o) {    // 2 switch
+            case Integer i -> {}
+            case Point(int x, int y) when (x == 0 && y == 0) -> {} // 3, when - 4, &&
+            case Point(int x, int y) when x == 0 -> {} // 5, when
+            default -> {}
+        }
+    }
+
+    // violation below, 'Cyclomatic Complexity is 5 (max allowed is 0)'
+    void testSwitch3(Object o) {  // 1, function declaration
+        switch (o) {    // 2 switch
+            case Integer i -> {}
+            case Point(int x, int y) -> {
+                if (x == 0 && y == 0) { // 3, if - 4, &&
+
+                }
+            }
+            case Point2(int x) -> {
+                if (x == 0) { // 5 if
+
+                }
+            }
+            default -> {}
+        }
+    }
+
+    record Point(int x, int y) { }
+
+    record Point2(int x) { }
+}

--- a/src/xdocs/checks/metrics/cyclomaticcomplexity.xml
+++ b/src/xdocs/checks/metrics/cyclomaticcomplexity.xml
@@ -19,13 +19,21 @@
           static initializers and instance initializers</a>.
         </p>
         <p>
-          The complexity is equal to the number of decision points <code> + 1</code>.
-          Decision points: <code>if</code>, <code>while</code>
-          , <code>do</code>, <code>for</code>, <code>?:</code>, <code>catch</code>
-          , <code>switch</code>, <code>case</code>
-          statements and operators <code>&amp;&amp;</code> and <code>||</code>
-          in the body of target.
+          The complexity is equal to the number of decision points <code>+ 1</code>.
+          Decision points:
         </p>
+        <ul>
+          <li>
+            <code>if</code>, <code>while</code>, <code>do</code>, <code>for</code>,
+            <code>?:</code>, <code>catch</code>, <code>switch</code>, <code>case</code> statements.
+          </li>
+          <li>
+            Operators <code>&amp;&amp;</code> and <code>||</code> in the body of target.
+          </li>
+          <li>
+            <code>when</code> expression in case labels, also known as guards.
+          </li>
+        </ul>
         <p>
           By pure theory level 1-4 is considered easy to test, 5-7 OK, 8-10
           consider re-factoring to ease testing, and 11+ re-factor now as testing will be painful.
@@ -90,6 +98,8 @@
                     LAND</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LOR">
                     LOR</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_WHEN">
+                    LITERAL_WHEN</a>
                   .
               </td>
               <td>
@@ -113,6 +123,8 @@
                     LAND</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LOR">
                     LOR</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_WHEN">
+                    LITERAL_WHEN</a>
                   .
               </td>
               <td>3.2</td>

--- a/src/xdocs/checks/metrics/cyclomaticcomplexity.xml.template
+++ b/src/xdocs/checks/metrics/cyclomaticcomplexity.xml.template
@@ -19,13 +19,21 @@
           static initializers and instance initializers</a>.
         </p>
         <p>
-          The complexity is equal to the number of decision points <code> + 1</code>.
-          Decision points: <code>if</code>, <code>while</code>
-          , <code>do</code>, <code>for</code>, <code>?:</code>, <code>catch</code>
-          , <code>switch</code>, <code>case</code>
-          statements and operators <code>&amp;&amp;</code> and <code>||</code>
-          in the body of target.
+          The complexity is equal to the number of decision points <code>+ 1</code>.
+          Decision points:
         </p>
+        <ul>
+          <li>
+            <code>if</code>, <code>while</code>, <code>do</code>, <code>for</code>,
+            <code>?:</code>, <code>catch</code>, <code>switch</code>, <code>case</code> statements.
+          </li>
+          <li>
+            Operators <code>&amp;&amp;</code> and <code>||</code> in the body of target.
+          </li>
+          <li>
+            <code>when</code> expression in case labels, also known as guards.
+          </li>
+        </ul>
         <p>
           By pure theory level 1-4 is considered easy to test, 5-7 OK, 8-10
           consider re-factoring to ease testing, and 11+ re-factor now as testing will be painful.

